File: QID-952-SFEGBMProcess/SFEGBMProcess.m
clear all
close all
clc

% user inputs parameters
disp('Please input method type as: 1 - Direct Integration, 2 - Euler Scheme') ;
disp(' ') ;
method = input('Input method ');
disp(' ') ;

% set parameters
n     = 1;
x0    = 0.84;
mu    = 0.02;
sigma = sqrt(0.1);
delta = 1 / 1000;

% main computation
t     = (0:ceil(n / delta));
sizet = size(t);
for i = 1:sizet(1, 2)
    aux(i) = 1;
end
x    = aux';
x(1) = x0;
no   = normrnd(0, 1, sizet(1, 2) - 1, 1) * sqrt(delta);

if method == 1 
    x = x0 * exp(cumsum((mu - 0.5 * sigma ^ 2) * delta + sigma .* no));
    x = [x0; x];
else
    i = 2;
    while (i <= sizet(1, 2))
        x(i) = x(i - 1) + mu * x(i - 1) * delta + sigma * x(i - 1) * no(i - 1);
        i    = i + 1;
    end
end

% output
plot(x)
title('Geometric Brownian Motion')
xlim([0 1000])

File: QID-952-SFEGBMProcess/SFEGBMProcess.R

# clear variables and close windows
rm(list = ls(all = TRUE))
graphics.off()

print("Please input method type as: 1 - Direct Integration, 2 - Euler Scheme")
method = 1  # if you want to apply Euler Scheme take 2 instead of 1

n      = 1
x_0    = 0.84
mu     = 0.02
sigma  = sqrt(0.1)
delta  = 1/1000
t      = c(0:ceiling(n/delta))
dimt   = dim(t(t))
x      = array(1, dim = dimt)
x[1]   = x_0
no     = rnorm((dimt[2] - 1), mean = 0, sd = 1) * sqrt(delta)

if (method == 1) {
    x = x_0 * exp(cumsum((mu - 0.5 * sigma^2) * delta + sigma * no))
}
{
    for (i in 2:dimt[2]) {
        x[i] = x[i - 1] + mu * x[i - 1] * delta + sigma * x[i - 1] * no[i - 1]
    }
}

if (method == 1) {
    x = c(x_0, x)
}
x = c(t(x))

# plot
par(mfrow = c(1, 1))
margins = c(1, 1, 1, 1)
plot(x, type = "l", col = "blue", main = "Geometric Brownian Motion", xlab = "t", 
    ylab = expression(S[t]), xlim = c(0, 1000))
